crates/suins-indexer/src/indexer.rs

```diff
@@ -0,0 +1,219 @@
+                sub_domain_wrappers.insert(
+                    sub_domain.nft.domain_name,
+                    sub_domain.id.id.bytes.to_string(),
+                );
+            };
+        }
+    }
+
+    /// For each input object, we're parsing the name record deletions
+    /// A deletion we want to track is a deleted object which is of `NameRecord` type.
+    /// Domain replacements do not count as deletions, but instead are an update to the latest state.
+    pub fn parse_name_record_deletions(
+        &self,
+        checkpoint: &CheckpointData,
+        removals: &mut Vec<String>,
+    ) {
+        // Gather all object ids that got deleted.
+        // This way, we can delete removed name records
+        // (detects burning of expired names or leaf names removal).
+        let deleted_objects: Vec<_> = checkpoint
+            .transactions
+            .iter()
+            .flat_map(|x| x.effects.all_removed_objects())
+            .map(|((id, _, _), _)| id)
+            .collect();
+
+        for input in checkpoint.input_objects() {
+            if self.is_name_record(input) && deleted_objects.contains(&input.id()) {
+                removals.push(input.id().to_string());
+            }
+        }
+    }
+
+    /// Processes a checkpoint and produces a list of `updates` and a list of `removals`
+    ///
+    /// We can then use these to execute our DB bulk insertions + bulk deletions.
+    ///
+    /// Returns
+    /// - `Vec<VerifiedDomain>`: A list of NameRecord updates for the database (including sequence number)
+    /// - `Vec<String>`: A list of IDs to be deleted from the database (`field_id` is the matching column)
+    pub fn process_checkpoint(
+        &self,
+        checkpoint: CheckpointData,
+    ) -> (Vec<VerifiedDomain>, Vec<String>) {
+        let mut name_records: Vec<NameRecordChange> = vec![];
+        let mut subdomain_wrappers: HashMap<String, String> = HashMap::new();
+        let mut removals: Vec<String> = vec![];
+
+        self.parse_name_record_changes(
+            &checkpoint.output_objects(),
+            &mut name_records,
+            &mut subdomain_wrappers,
+        );
+
+        self.parse_name_record_deletions(&checkpoint, &mut removals);
+
+        let updates = prepare_db_updates(
+            &name_records,
+            &subdomain_wrappers,
+            checkpoint.checkpoint_summary.sequence_number,
+        );
+
+        (updates, removals)
+    }
+}
+
+/// Allows us to format a SuiNS specific query for updating the DB entries
+/// only if the checkpoint is newer than the last checkpoint we have in the DB.
+/// Doing that, we do not care about the order of execution and we can use multiple threads
+/// to commit from later checkpoints to the DB.
+pub fn format_update_field_query(field: &str) -> String {
+    format!(
+        "CASE WHEN excluded.last_checkpoint_updated > domains.last_checkpoint_updated THEN excluded.{field} ELSE domains.{field} END"
+    )
```

Comment:
For this and other SQL queries we don't plan on these being seeded with user provided input correct? There's always a risk with string interpolation and SQL injection so ideally we'd use the diesel DSL to construct this, if possible, to remove the possibility of this. But if the provided "field" value is always controlled by us then it wouldn't be an issue to leave this as is.

Reply:
Looking further below it seems we do fully control inputs here so we're good

Reply:
Correct, there are no external inputs so I am not worried about this. But, I'll add a "warning" comment about this to make sure anyone using/looking at the code in the future sees it.